pkg/reconciler/channel/channel.go

```diff
@@ -0,0 +1,412 @@
+				channel.Status.MarkTopicReady()
+			} else if c.IsUnknown() {
+				channel.Status.MarkTopicOperating(c.Reason, c.Message)
+			} else if c.IsFalse() {
+				channel.Status.MarkNoTopic(c.Reason, c.Message)
+			}
+		}
+	}
+	// 2. Sync all subscriptions.
+	//   a. create all subscriptions that are in spec and not in status.
+	//   b. delete all subscriptions that are in status but not in spec.
+	if err := r.syncSubscribers(ctx, channel); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *Reconciler) updateStatus(ctx context.Context, desired *v1alpha1.Channel) (*v1alpha1.Channel, error) {
+	channel, err := c.channelLister.Channels(desired.Namespace).Get(desired.Name)
+	if err != nil {
+		return nil, err
+	}
+	// If there's nothing to update, just return.
+	if reflect.DeepEqual(channel.Status, desired.Status) {
+		return channel, nil
+	}
+	becomesReady := desired.Status.IsReady() && !channel.Status.IsReady()
+	// Don't modify the informers copy.
+	existing := channel.DeepCopy()
+	existing.Status = desired.Status
+
+	ch, err := c.RunClientSet.EventsV1alpha1().Channels(desired.Namespace).UpdateStatus(existing)
+	if err == nil && becomesReady {
+		duration := time.Since(ch.ObjectMeta.CreationTimestamp.Time)
+		c.Logger.Infof("Channel %q became ready after %v", channel.Name, duration)
+
+		if err := c.StatsReporter.ReportReady("Channel", channel.Namespace, channel.Name, duration); err != nil {
+			logging.FromContext(ctx).Infof("failed to record ready for Channel, %v", err)
+		}
+	}
+
+	return ch, err
+}
+
+func (c *Reconciler) syncSubscribers(ctx context.Context, channel *v1alpha1.Channel) error {
+	if channel.Status.Subscribers == nil {
+		channel.Status.Subscribers = []eventingduck.SubscriberStatus(nil)
+	}
+
+	subCreates := []eventingduck.SubscriberSpec(nil)
+	subUpdates := []eventingduck.SubscriberSpec(nil)
+	subDeletes := []eventingduck.SubscriberStatus(nil)
+
+	// Make a map of name to PullSubscription for lookup.
+	pullsubs := make(map[string]pubsubv1alpha1.PullSubscription)
+	if subs, err := c.getPullSubscriptions(ctx, channel); err != nil {
+		c.Logger.Infof("Failed to list PullSubscriptions, %s", err)
+	} else {
+		for _, s := range subs {
+			pullsubs[s.Name] = s
+		}
+	}
+
+	exists := make(map[types.UID]eventingduck.SubscriberStatus)
+	for _, s := range channel.Status.Subscribers {
+		exists[s.UID] = s
+	}
+
+	if channel.Spec.Subscribable != nil {
+		for _, want := range channel.Spec.Subscribable.Subscribers {
+			if got, ok := exists[want.UID]; !ok {
+				// If it does not exist, then update it.
+				subCreates = append(subCreates, want)
+			} else {
+				_, found := pullsubs[resources.GenerateSubscriptionName(want.UID)]
+				// If did not find or the PS has updated generation, update it.
+				if !found || got.ObservedGeneration != want.Generation {
+					subUpdates = append(subUpdates, want)
+				}
+			}
+			// Remove want from exists.
+			delete(exists, want.UID)
+		}
+	}
+
+	// Remaining exists will be deleted.
```

Comment:
This is a pretty long function, took me a while to get through and figure out under which cases what happens. Might want to include more comments here as well as what the overall method does. Also might consider breaking add/update/delete loops into their own methods?

Reply:
#113 will do this.